File: solution-python/033_search_in_rotated_sorted_array.py
"""
	033. Search in Rotated Sorted Array
	@name: li jin
	@date: Apr 21, 2016
	@link: https://leetcode.com/problems/search-in-rotated-sorted-array/
	@time: 44 ms
"""

class Solution(object):
    def search(self, nums, target):
        if not nums:
            return -1
        start = 0; end = len(nums)-1
        while start + 1 < end:
            mid = (start+end)/2
            if nums[mid] == target:
                return mid
            if nums[start] <= target < nums[mid]:
                end = mid
            elif nums[mid] < target <= nums[end]:
                start = mid
            elif nums[mid] > nums[end]: # right side is pivoted
                start = mid
            else:
                end = mid
        if nums[start] == target:return start
        elif nums[end] == target:return end
        return -1

File: solution-cpp/033_search_in_rotated_sorted_array.cpp
/*
	033. Search in Rotated Sorted Array
	@name: li jin
	@date: Apr 21, 2016
	@link: https://leetcode.com/problems/search-in-rotated-sorted-array/
	@time: 4 ms
*/

class Solution {
public:
    int search_helper(vector<int>& nums, int target, int low, int high) {
        if (low > high) return -1;
        if (low == high) return target == nums[low] ? low : -1;
        int mid = (high + low) >> 1;
        if (target == nums[mid]) return mid;
        if (target > nums[mid]) {
            if (nums[high] > target) return search_helper(nums, target, mid+1, high);
            int tmp = search_helper(nums, target, mid+1, high);
            if (tmp == -1) return search_helper(nums, target, low, mid-1);
            else return tmp;
        } else {
            if (nums[low] < target) return search_helper(nums, target, low, mid-1);
            int tmp = search_helper(nums, target, low, mid-1);
            if (tmp == -1) return search_helper(nums, target, mid+1, high);
            else return tmp;
        }
    }

    int search(vector<int>& nums, int target) {
        return search_helper(nums, target, 0, nums.size()-1);
    }
};
